lib/sycamore/sycamore/document.py

```diff
@@ -0,0 +1,390 @@
+
+    def accept(self, visitor):
+        return visitor.visit_section_header(self)
+
+
+class Table(Leaf):
+    def __init__(self, node_id, box, content, properties, continued=None):
+        super().__init__(node_id, box, content, properties)
+
+    def category(self):
+        return Category.Caption
+
+    def accept(self, visitor):
+        return visitor.visit_table(self)
+
+
+class Text(Leaf):
+    def __init__(self, node_id, box, content, properties, continued=None):
+        super().__init__(node_id, box, content, properties)
+        self._continued = continued
+
+    def category(self):
+        return Category.Caption
+
+    def continued(self):
+        return self._continued
+
+    def accept(self, visitor):
+        return visitor.visit_text(self)
+
+
+class Title(Leaf):
+    def __init__(self, node_id, box, content, properties):
+        super().__init__(node_id, box, content, properties)
+
+    def category(self):
+        return Category.Caption
+
+    def accept(self, visitor):
+        return visitor.visit_title(self)
+
+
+class Group(Internal):
+    """
+    Group semantic related objects together, e.g. list items, figure and caption
+    """
+
+    def __init__(self, node_id, children, properties):
+        super().__init__(node_id, children, properties)
+
+    def category(self):
+        return Category.Group
+
+    def accept(self, visitor):
+        return visitor.visit_group(self)
+
+
+class Section(Internal):
+    def __init__(self, node_id, children, header, properties):
+        super().__init__(node_id, children, properties)
+        self._header = header
+
+    def category(self):
+        return Category.Section
+
+    def accept(self, visitor):
+        return visitor.visit_section(self)
+
+
+class Visitor(ABC):
+    @abstractmethod
+    def visit_caption(self, caption: Caption):
+        pass
+
+    @abstractmethod
+    def visit_footnote(self, footnote: Footnote):
+        pass
+
+    @abstractmethod
+    def visit_formula(self, formula: Formula):
+        pass
+
+    @abstractmethod
+    def visit_list_item(self, list_item: ListItem):
+        pass
+
+    @abstractmethod
+    def visit_page_footer(self, page_footer: PageFooter):
+        pass
+
+    @abstractmethod
+    def visit_page_header(self, page_header: PageHeader):
+        pass
+
+    @abstractmethod
+    def visit_picture(self, picture: Picture):
+        pass
+
+    @abstractmethod
+    def visit_section_header(self, section_header: SectionHeader):
+        pass
+
+    @abstractmethod
+    def visit_table(self, table: Table):
+        pass
+
+    @abstractmethod
+    def visit_text(self, text: Text):
+        pass
+
+    @abstractmethod
+    def visit_title(self, title: Title):
+        pass
+
+    @abstractmethod
+    def visit_group(self, group: Group):
+        pass
+
+    @abstractmethod
+    def visit_section(self, section: Section):
+        pass
+
+
+class NaiveSemanticVisitor(Visitor):
+    def visit_caption(self, caption):
```

Comment:
Hey, thanks for your feedback. I actually have different opinions on this.
1. this does not align with a pre-existing design pattern
2. Considering code scalability, I feel there is more requirement to extend new functions/transformations instead of a new node type.
3. the visitor considers more about the inheritance, for the above naivesemanticvisitor example, it's possible to have requirement for a newer semanticvisitor which only modifies the function for handling table and leave all the other unchanged.